exploration/number-selection.md

```diff
@@ -0,0 +1,165 @@
+
+MF1 supported selection on either cardinal plurals or ordinal numbers,
+via the `plural` and `selectordinal` selectors.
+It also customized this selection beyond the capabilities of `com.ibm.icu.text.PluralRules`
+by allowing for explicit value matching and an `offset` parameter.
+
+As pointed out by <a href="https://github.com/mihnita">@mihnita</a> in particular,
+category selection is not always appropriate for selection on a number:
+the number may be representing some completely other quantity,
+such as a four-digit year or the integer value of an enumerator.
+
+Furthermore, as pointed out by <a href="https://github.com/ryzokuken">@ryzokuken</a>
+in <a href="https://github.com/unicode-org/message-format-wg/pull/457#discussion_r1307443288">#457 (comment)</a>,
+ordinal selection is similar to plural selection, but not identical.
+
+Additionally,
+MF1 provides `ChoiceFormat` selection based on whether a number is within a range,
+and both JS and ICU PluralRules implementations provide for determining the plural category
+of a range based on its start and end values.
+These range-based selectors are not initially considered here.
+
+## Use-Cases
+
+Put together, we have thus identified three different types of selection
+we would like to support on numeric values:
+
+- cardinal plural selection
+- ordinal selection
+- exact value match selection
+
+## Requirements
+
+- Enable cardinal plural selection.
+- Enable ordinal number selection.
+- Enable exact match selection.
+- Support relevant formatting options, such as `minimumFractionDigits`.
+- Encourage developers to provide the formatting options used in patterns to the selector
+  so that proper selection can be done.
+
+## Constraints
+
+ICU MF1 messages using `plural` and `selectordinal` should be representable in MF2.
+
+## Proposed Design
+
+Given that we already have a `:number`,
+it makes sense to add a `<matchSignature>` to it with an option
+
+```xml
+<option name="select" values="plural ordinal exact" default="plural" />
+```
+
+The default `plural` value is presumed to be the most common use case,
+and it affords the least bad fallback when used incorrectly:
+Using "plural" for "exact" still selects exactly matching cases,
+whereas using "exact" for "plural" will not select LDML category matches.
+This might not be noticeable in the source language,
+but may cause problems in target languages that the original developer is not considering.
+
+Additional options such as `minimumFractionDigits` and others already supported by `:number`
+should also be supported.
+
+If PR [#532](https://github.com/unicode-org/message-format-wg/pull/532) is accepted,
+also add the following `<alias>` definitions to `<function name="number">`:
+
+```xml
+<alias name="plural" supports="match">
+	<setOption name="select" value="plural"/>
+</alias>
+<alias name="ordinal" supports="match">
+	<setOption name="select" value="ordinal"/>
+</alias>
+```
```

Comment:
Is this list exhaustive? #532 includes `:integer`. Presumably `:integer` is an integer plural selector by default (in addition to being a formatter).

Other candidates for shorthands might be `:percent`, `:currency` and (maybe??) `:scientific`. The first two are shorthands in ICU4J MF1. `scientific` is a numeric formatting option (but not exposed by MF1). 

There is also the potentially lamentable `:spellout` from ICU4J MF1.

I'm guessing that what we want to promote is single-annotation messages, e.g.:

```
.input {$var :spellout}
.match {$var}                               // it's already a plural
[0]   {{You have no more bone dragons}}
[one] {{You have {$var} more bone dragon}}  // "You have one more..."
[*]   {{You have {$var} more bone dragons}}
```

Reply:
I think the ones you mention are all formatters? The list here is purely about `.match` selectors, and for that I think just `:plural` and `:ordinal` could be enough. If you need exact matching, then the relatively rare `:number select=exact` is pretty good?

Reply:
To clarify a bit, a proposed alias like `:integer` or `:spellout` still ends up calling the `:number` function, and so if the expression doesn't have an explicit `select` attribute and the alias does not set `supports="format"`, it'll end up inheriting the default `select="plural"` for selection, together with the rest of the expression & alias options.

Reply:
Here's how I would classify the different aliases that have been mentioned so far:
- `:integer` -- The most obvious one, and therefore included already in [#532](https://github.com/unicode-org/message-format-wg/pull/532).
- `:plural` and `:ordinal` -- Functions for selection only with `supports="match"`. Included in this design doc.
- `:percent`, `:currency`,`:scientific`, `:spellout` -- Possible aliases for later/separate consideration. As with `:integer`, they might be used also for selection, but I don't see how any of them would specify a `<setOption name="select">`. So I don't think they belong in a design doc on Number Selection.

Once #532 lands, would a PR adding that third set be the right focal point for discussion of their merits, rather than this PR?

Reply:
I agree that `:ordinal` could also be a formatter, but I am not here proposing that it be one of the core formatters that we expect all MF2 implementations to provide. That could be considered as a separate further step, should this design on number _selection_ first introduce it as a selector.

Reply:
I would very strongly prefer taking this smaller step first, and not expanding the scope of this design document.